csharp/ql/lib/semmle/code/csharp/security/dataflow/ZipSlipQuery.qll

```diff
@@ -204,6 +204,32 @@ private module SanitizedGuardTaintTrackingConfiguration implements DataFlow::Con
   }
 }
 
+/**
+ * A Callable that successfully validates a path will resolve under a given directory,
+ * and if it does not, throws an exception.
+ */
+private class ValidatingCallableThrowing extends Callable{
+  Parameter paramFilename;
+  ValidatingCallableThrowing(){
+    paramFilename = this.getAParameter() and
+    // It passes the guard, contraining the function argument to the Guard argument.
+    exists(ZipSlipGuard g, DataFlow::ParameterNode source, DataFlow::Node sink |
+      g.getEnclosingCallable() = this and
+      source = DataFlow::parameterNode(paramFilename) and
+      sink = DataFlow::exprNode(g.getFilePathArgument()) and
+      SanitizedGuardTT::flow(source, sink) and
+      exists(AbstractValues::BooleanValue bv, ThrowStmt throw |
+        throw.getEnclosingCallable() = this and
+        forall(TryStmt try | try.getEnclosingCallable() = this | not throw.getParent+() = try) and
+        // If there exists a control block that guards against misuse
+        bv.getValue() = false and
+        g.controlsNode(throw.getAControlFlowNode(), bv)
+      )
+    )
```

Comment:
So this is only checking one level. So if you had a method which validated a path by forwarding to another method then this wouldn't work, right? It's comforting to know that we could easily make this recursive if we start getting FP reports on more complicated interprocedural barriers.

Sidenote: I know GitHub has recently done some work on making this Just Work when you define a barrier guard. [Here's the PR](https://github.com/github/codeql/pull/20183) that does it for Java, and I hope that the C# one is also coming soon 🤞 That would make these complexities unnecessary.